### ISSUE TITLE
simple extension for geometry metadata

### ISSUE THREAD

**Issue by zellski** [KhronosGroup/glTF#1478]
Hey folks,

Throwing together a simple internal extension to allow glTF to summarise some computed geometry data –– I thought I might as well post here in case anyone else is interested.

The information we are interested in is **vertex count**, **triangle count**, and **world-space bounds** for each mesh. Since many nodes can instantiate a mesh, nodes are the meaningful place to insert the extension data.

Every node includes data on the cumulative counts and world bounds for the subtree they parent. If the node references a mesh, we also include the world bounds of just that node.

Limitations:
- This is just for the default static pose. We make no attempt to accomodate animations.
- We traverse the scene graph from the default scene. Other scenes would typically result in different per-node world transforms.

Proposed output:
```
{
    "asset": { "version": "2.0" },
    "extensionsUsed": [ "FB_geometry_metadata" ],
    "scenes": [
        {"nodes": [ 0 ]},
        {"nodes": [ 2 ]}
    ],
    "nodes": [
        {
            "children": [ 1 ],
            "extensions": {
                "FB_geometry_metadata": {
                    "nodeWorldBounds": {
                        "min": [ -0.26, 0.24, -0.88 ],
                        "max": [ 2.24, 1.88, -0.22 ]
                    },
                    "vertexCount": 48,
                    "triangleCount": 24
                }
            }
        },
        {
            "mesh": 0,
            "children": [ 2 ],
            "extensions": {
                "FB_geometry_metadata": {
                    "meshWorldBounds": {
                        "min": [ -0.13, 0.12, -0.44 ],
                        "max": [ 1.12, 0.94, -0.11 ]
                    },
                    "nodeWorldBounds": {
                        "min": [ -0.26, 0.24, -0.88 ],
                        "max": [ 2.24, 1.88, -0.22 ]
                    },
                    "vertexCount": 48,
                    "triangleCount": 24
                }
            }
        },
        {
            "scale": [ 2.0, 2.0, 2.0 ],
            "mesh": 0,
            "extensions": {
                "FB_geometry_metadata": {
                    "meshWorldBounds": {
                        "min": [ -0.26, 0.24, -0.88 ],
                        "max": [ 2.24, 1.88, -0.22 ]
                    },
                    "nodeWorldBounds": {
                        "min": [ -0.26, 0.24, -0.88 ],
                        "max": [ 2.24, 1.88, -0.22 ]
                    }
                    "vertexCount": 24,
                    "triangleCount": 12
                }
            }
        }
    ],
    "meshes": [
        {
            "primitives": [
                {
                    "indices": 0,
                    "attributes": { "POSITION": 1 }
                }
            ],
        }
    ],
    ... rest of file
}
```

**Comment by lexaknyazev** [comment]
> This is just for the default static pose.

1. Does this extension take into account initial morph state (stored in `mesh.primitive.targets`)?
2. Same question wrt skins.

**Comment by zellski** [comment]
> > This is just for the default static pose.
> 
> 1. Does this extension take into account initial morph state (stored in `mesh.primitive.targets`)?
> 2. Same question wrt skins.

Hmm. That would probably be the "least astonishing" behaviour, yeah. On the other hand, it does  raise the complexity bar in generating the output. Only quite full-featured pipelines and generators would be able to do it.

I'm torn. What do you think?

**Comment by lexaknyazev** [comment]
First of all, an extension spec should document its limitations in either case. And since it's a vendor extension, it could do whatever the vendor sees fit.

In case this was a multi-vendor or Khronos extension, I'd say that usefulness of bounding boxes depends on application-specific usage patterns and various helper data could be included on a case by case basis, i.e., an extension could have several optional fields like: "no-transform", "static-transform", etc.

**Comment by donmccurdy** [comment]
This feels like a scenario where a common library would be as useful as the extension itself — all of this information should be cheap to compute (assuming static transforms) even at runtime, it's just non-trivial to write the code to do so.

**Comment by donmccurdy** [comment]
I've changed my mind about my comment above; seems that accessor min/max is insufficient to compute some of this. Using extension makes sense to me, with some possibility of moving this into the core format in the future.

**Comment by zellski** [comment]
> I've changed my mind about my comment above; seems that accessor min/max is insufficient to compute some of this. Using extension makes sense to me, with some possibility of moving this into the core format in the future.

Yeah, I think min/max can yield estimated bounding boxes anywhere from perfect to wildly wrong, depending. Maybe there's some cleverness that escapes me, but I'm not sure you can avoid essentially `worldMatrix` × `vertexPosition` for every vertex. It gets pretty expensive.

At least up to the `static-transform` case, the algorithm is pretty straight-forward.

**Comment by elalish** [comment]
I too would find this extension useful, as calculating these bounds at runtime can be quite costly. However, for camera framing I find bounding boxes to be not very helpful, as they don't fit well in a perspective view frustum. Could we consider something more general that would allow us to calculate our own framing? 

I'm thinking of a basic hull: just a set of vertices (no mesh) that an author can use to represent the framed volume of the scene; ideally with few enough vertices to give a decent approximation without much runtime computation load. We could also specify a center point so that we don't have to assume the model should be centered at the middle of its bounding box. 

The framing volume doesn't have to be "tight", nor does it need to enclose all geometry, as it can simply represent the artist's focus, perhaps leaving background geometry outside or containing enough space to enclose all of the animations, as desired.

**Comment by atteneder** [comment]
A custom hull vertex list, considering animation or even per-node bounds might be overkill though

Most common case is probably to place/scale an entire scene correctly and a per scene bouding box would be sufficient for that
